### PR TITLE
[MERGE][REF] base, *: Compute fields before creation.

### PR DESCRIPTION
**Pull request by Feyensv**
Currently, the computed fields are computed after record creation in DB.  This means that computed fields cannot be required and that no SQL constraints can be based on those fields.

To support required and sql constraints on computed (and stored) fields, we need to compute the fields before record creation ...

##### `_pre_compute` model attribute

Specify a model as `_pre_compute=True` to enable computed fields computation before record creation.  Added to make the feature OPT-IN, allowing us to convert models one by one when there is a gain to pre_compute their computed fields.

##### `pre_compute` field attribute

Specify fields as `pre_compute=False` to ensure they are computed after record creation

* Done by default if any compute dependency is `pre_compute=False`
* Done by default if field depends on `create_date/write_date/create_uid/write_uid`
* Done manually for part of existing fields (mainly statistics fields)

##### Postpone NOT NULL AND SQL constraints after fields/models computation/setup

Column are set to NOT NULL after fields computation in DB to ensure model extension with required computed fields works fine.

##### New Record cannot impact real records And inversely

To pre-compute computed stored fields, we instantiate a new record with the values given to the create call.  This meant that two records were put in cache for one creation (one real and one new).
When a record is created linking to real records in o2m or m2m fields, the inverse was triggered and you could have new records in the relational fields of real records.

Example:
Create an `account.move.line`, linking to real `account.move` 1, if you accessed the lines from the `account.move`, you would receive 2 records (one new and one real).  

To ensure the new `new()` call in the `create()` method doesn't impact real records, the inverse of new (resp. real) records changes is not applied on real (resp. new) records anymore.  It was already the strategy of the orm, but is more strictly applied from now on.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/models.py [hunk 1]
```diff
@@ -71,6 +71,7 @@
 regex_field_agg = re.compile(r'(\w+)(?::(\w+)(?:\((\w+)\))?)?')
 
 AUTOINIT_RECALCULATE_STORED_FIELDS = 1000
+PROTECTED_FIELDS = ["id", "create_date", "write_date", "create_uid", "write_uid"]
```
Comment (Feyensv) [inline odoo/models.py]: to remove

File: addons/survey/models/badge.py [hunk 2]
```diff
@@ -8,9 +8,10 @@ class GamificationBadge(models.Model):
     _inherit = 'gamification.badge'
 
     survey_ids = fields.One2many('survey.survey', 'certification_badge_id', 'Survey Ids')
-    survey_id = fields.Many2one('survey.survey', 'Survey', compute='_compute_survey_id', store=True)
+    survey_id = fields.Many2one(
+        'survey.survey', 'Survey', compute='_compute_survey_id', store=True, pre_compute=False)
 
-    @api.depends('survey_ids.certification_badge_id')
+    @api.depends('survey_ids')
```
Comment (Feyensv) [inline addons/survey/models/badge.py]: TODO check with rco

File: addons/survey/models/badge.py [hunk 3]
```diff
@@ -8,9 +8,10 @@ class GamificationBadge(models.Model):
     _inherit = 'gamification.badge'
 
     survey_ids = fields.One2many('survey.survey', 'certification_badge_id', 'Survey Ids')
-    survey_id = fields.Many2one('survey.survey', 'Survey', compute='_compute_survey_id', store=True)
+    survey_id = fields.Many2one(
+        'survey.survey', 'Survey', compute='_compute_survey_id', store=True, pre_compute=False)
 
-    @api.depends('survey_ids.certification_badge_id')
```
Comment (Feyensv) [inline addons/survey/models/badge.py]: check with rco

File: addons/account/models/account_bank_statement.py [hunk 4]
```diff
@@ -484,6 +486,7 @@ def _check_currency_id(self):
     @api.model
     def create(self, vals):
         line = super(AccountBankStatementLine, self).create(vals)
+        # VFE FIXME wouldn't this be fixed with journal_currency_id stored ?
```
Comment (Feyensv) [inline addons/account/models/account_bank_statement.py]: check with rco

Comment (Feyensv) [inline addons/account/models/account_bank_statement.py]: check with rco

**Comment by Feyensv** [comment]
@tde-banana-odoo If you wanna have a look ;).
Application on `sale` scope works fine and can be found in #42709 (the red runbot comes from my stupid perf tests logging a little warning :D).

File: odoo/fields.py [hunk 5]
```diff
@@ -2821,11 +2846,18 @@ def convert_to_write(self, value, record):
...
+                        values = dict()
+                        for fname in record._cache:
+                            if fname in inv_names:
+                                continue
+                            if record[fname] == origin[fname]:
+                                continue
+                            field = record._fields[fname]
+                            val = field.convert_to_write(record[fname], record)
+                            if not record.id and field.type in ("one2many", "many2many")\
+                                and len(val) == 1 and val[0][0] == 6 and val[0][2] == origin[fname].ids:
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/fields.py]: It would be easier if `convert_to_write()` returns `[]` when nothing has changed in the x2many...

File: odoo/models.py [hunk 6]
```diff
@@ -2509,12 +2511,18 @@ def _auto_init(self):
...
+                        field = recs._fields[fname]
+                        self.env.add_to_compute(field, recs)
+                        if not must_create_table and field.required and field.column_type:
+                            # Required computed fields columns have to be
+                            # set NOT NULL after field computation
+                            # when the table already exists (IntegrityError)
+                            field.update_db_notnull(self, columns.get(field.name))
+
+            if self._auto:
+                self.pool.post_init(self._add_sql_constraints)
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply odoo/models.py]: Done before adt PR, I'll remove this ;).

File: odoo/fields.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Feyensv) [reply odoo/fields.py]: No, because it means we would put an empty `[ ]` in `values[fname]`, an a `(1, 0, {fname:[ ]}` would still be added to the returned result, bringing back the same problem of writing for no effective change.

I added a test to cover the case.

File: odoo/addons/test_new_api/models/test_new_api.py [hunk 7]
```diff
@@ -856,3 +857,66 @@ class StateMixin(models.AbstractModel):
...
+    duplicates = fields.Many2many(
+        comodel_name='test_new_api.model_advanced_computes',
+        relation='test_new_api_advanced_computes_rel', column1="rec1", column2="rec2",
+        compute="_compute_duplicates", pre_compute=False, store=True)
+
+    @api.depends('name1', 'name2')
+    def _compute_uppers(self):
+        for rec in self:
+            if rec.id or not rec.env.context.get('creation', False):
+                raise ValidationError("Should be computed before record creation")
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply odoo/addons/test_new_api/models/test_new_api.py]: Currently, it guarantees that the field is computed before DB insert and not (re)computed after.
I want to ensure the computes are only done once by create (and before/after depending on `pre_compute` attr)

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 8]
```diff
@@ -2457,8 +2477,97 @@ class TestSelectionDeleteUpdate(common.TransactionCase):
...
+class TestPrePostComputes(common.TransactionCase):
+
+    def test_pre_post_create_computes(self):
+        Model = self.env["test_new_api.model_advanced_computes"]
+
+        # Ensure automatic assignation of pre_compute=False works
+        # 1) dependency on create_date/create_uid/write_date/write_uid
+        self.assertFalse(Model._fields.get('create_month').pre_compute)
+        # 2) dependency on a pre_compute=False field
+        self.assertFalse(Model._fields.get('full_card_content').pre_compute)
```
Comment (Feyensv) [inline odoo/addons/test_new_api/tests/test_new_fields.py]: To remove, since optin feature, or force pre_compute=True on the fields then check they are still put to pre_compute=False ?

File: odoo/fields.py [hunk 9]
```diff
@@ -199,6 +200,16 @@ def _read_group_many2one_field(self, records, domain, order):
 
         .. seealso:: :ref:`Advanced Fields/Compute fields <reference/fields/compute>`
 
+    :param bool pre_compute: whether the field should be computed before record insertion
+        in database.  Should be used to specify manually some fields as pre_compute=False
+        when the field HAS to be computed after record insertion.
+        (e.g. statistics fields based on search/read_group), many2one
+        linking to the previous record, ... (default: `True`)
```
Comment (Feyensv) [inline odoo/fields.py]: default: False

File: odoo/fields.py [hunk 10]
```diff
@@ -616,6 +627,9 @@ def resolve_depends(self, registry):
                 if field is self and index:
                     self.recursive = True
 
+                if not field.pre_compute or fname in POST_CREATE_FIELDS:
+                    self.pre_compute = False
+
```
Comment (Feyensv) [inline odoo/fields.py:632]: Explicit > Implicit : leave it to the dev to specify it as not precomputable.

File: odoo/fields.py [hunk 11]
```diff
@@ -2647,6 +2664,8 @@ def _update_inverses(self, records, value):
         corecord = self.convert_to_record(value, records)
         for invf in records._field_inverses[self]:
             valid_records = records.filtered_domain(invf.get_domain_list(corecord))
+            # do not mix new records and real records
+            valid_records = valid_records.filtered(lambda r: bool(r.id) == bool(corecord.id))
```
Comment (Feyensv) [inline odoo/fields.py:2668]: Can probably be removed, to check

File: odoo/modules/registry.py [hunk 12]
```diff
@@ -332,6 +332,9 @@ def post_init(self, func, *args, **kwargs):
 
     def post_constraint(self, func, *args, **kwargs):
         """ Call the given function, and delay it if it fails during an upgrade. """
+        if kwargs.pop("postpone", False):
+            self._constraint_queue.append(partial(func, *args, **kwargs))
+            return
```
Comment (Feyensv) [inline odoo/modules/registry.py:337]: Error log should be removed by adt in his own pr and then this hack could be removed.

File: odoo/models.py [hunk 13]
```diff
@@ -3667,6 +3692,97 @@ def _write(self, vals):
...
+            return
+
+        missing_fields = [
+            field for field in stored_computed_fields
+            if field.name not in vals
+        ]
+
+        if missing_fields:
+            # Specific context to know when we are in this specific case.
+            new_obj = self.with_context(create_pre_compute=True).new(vals)
```
Comment (Feyensv) [inline odoo/models.py:3714]: context isn't useful anymore ?

**Comment by rco-odoo** [comment]
@Feyensv I lack a clear spec of what should be done in each case.  I propose the following spec table, that lists all possible cases for a simple stored field (x2many need some extra specs).  The only change should be the effects in the last column.  Non-precomputed fields should behave as currently.

Given value | Field.default | Field.compute | Current effect | Expected effect
--- | --- | --- | --- | ---
none | none | none | none
value | none | none | INSERT value
none | default | none | INSERT default
value | default | none | INSERT value
none | none | compute readonly | no INSERT, post-compute | INSERT computed value
value | none | compute readonly | no INSERT, post-compute | INSERT computed value
none | default | compute readonly | no INSERT, post-compute | INSERT computed value
value | default | compute readonly | no INSERT, post-compute | INSERT computed value
none | none | compute read-write | no INSERT, post-compute | INSERT computed value
value | none | compute read-write | INSERT value
none | default | compute read-write | INSERT default
value | default | compute read-write | INSERT value
none | none | compute with inverse | no INSERT, post-compute | INSERT computed value
value | none | compute with inverse | INSERT value, then INVERSE field
none | default | compute with inverse | INSERT default, then INVERSE field
value | default | compute with inverse | INSERT value, then INVERSE field

The tests should be more systematic, and document each behavior.  I think that some of the current effects are not even correct at the moment.

File: addons/sms/wizard/sms_composer.py [hunk 14]
```diff
@@ -167,22 +169,24 @@ def _compute_body(self):
...
+    # def create(self, values):
+    #     # TDE FIXME: currently have to compute manually to avoid required issue, waiting VFE branch
+    #     if not values.get('body') or not values.get('composition_mode'):
+    #         values_wdef = self._add_missing_default_values([values])
+    #         cache_composer = self.new(values_wdef[0])
+    #         cache_composer._compute_body()
+    #         cache_composer._compute_composition_mode()
+    #         values['body'] = values.get('body') or cache_composer.body
+    #         values['composition_mode'] = values.get('composition_mode') or cache_composer.composition_mode
+    #     return super(SendSMS, self).create(values)
```
Comment (rco-odoo) [inline addons/sms/wizard/sms_composer.py:189]: Please delete instead of commenting out.

File: addons/event/models/event_event.py [hunk 15]
```diff
@@ -471,15 +473,17 @@ def name_get(self):
...
-        res = super(EventEvent, self).create(vals)
-        if res.organizer_id:
-            res.message_subscribe([res.organizer_id.id])
-        return res
+    @api.model_create_multi
+    def create(self, vals_list):
+        # for vals in vals_list:
+        #     # Temporary fix for ``seats_limited`` and ``date_tz`` required fields (see ``_compute_from_event_type``
+        #     vals.update(self._sync_required_computed(vals))
+
```
Comment (rco-odoo) [inline addons/event/models/event_event.py:481]: Please delete instead of commenting out.

File: addons/sms/wizard/sms_composer.py [hunk 16]
```diff
@@ -74,14 +75,15 @@ def default_get(self, fields):
         readonly=False, store=True, required=True)
 
     @api.depends('res_ids_count', 'active_domain_count')
-    @api.depends_context('default_composition_mode')
     def _compute_composition_mode(self):
         for composer in self:
-            if self.env.context.get('default_composition_mode') == 'guess' or not composer.composition_mode:
+            if self.env.context.get('composition_mode_default') == 'guess' or not composer.composition_mode:
```
Comment (rco-odoo) [inline addons/sms/wizard/sms_composer.py:80]: Is it safe to remove the `@api.depends_context()`?

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 17]
```diff
@@ -2610,11 +2630,28 @@ class TestSelectionDeleteUpdate(common.TransactionCase):
...
+                    JOIN ir_model_fields_selection selection on field_id=field.id
+                    WHERE selection.id=%i
+                )
+                UPDATE ir_model_fields field
+                SET state = 'manual'
+                FROM fields f
+                WHERE f.id=field.id
+            """% selections.id)
+            selections.invalidate_cache()
+        selections.unlink()
```
Comment (rco-odoo) [inline odoo/addons/test_new_api/tests/test_new_fields.py:2654]: This is noise, see #54004 for the right fix.

File: odoo/fields.py [hunk 18]
```diff
@@ -319,13 +326,15 @@ def _get_attrs(self, model, name):
             if not (attrs['store'] and not attrs.get('readonly', True)):
                 attrs['copy'] = attrs.get('copy', False)
             attrs['readonly'] = attrs.get('readonly', not attrs.get('inverse'))
+            attrs['pre_compute'] = attrs.get('pre_compute', model._pre_compute)
```
Comment (rco-odoo) [inline odoo/fields.py:329]: This makes unit testing more difficult.  Instead, add an explicit test on `model._pre_compute` in the auxiliary methods of `create`.  This flag must eventually disappear (to be discussed).

File: odoo/fields.py [hunk 19]
```diff
@@ -630,6 +644,10 @@ def resolve_depends(self, registry):
                 if field.type in ('one2many', 'many2many'):
                     for inv_field in Model._field_inverses[field]:
                         yield tuple(field_seq) + (inv_field,)
+                    if self.related:
+                        # related fields depending on x2m should always be post computed.
+                        # Otherwise, we could create x2m records multiple times wrongly
```
Comment (rco-odoo) [inline odoo/fields.py:649]: I don't understand the reason for "we could create x2m records multiple times".

File: odoo/fields.py [hunk 20]
```diff
@@ -2883,11 +2903,18 @@ def convert_to_write(self, value, record):
...
+                        for fname in record._cache:
+                            if fname in inv_names:
+                                continue
+                            if record[fname] == origin[fname]:
+                                continue
+                            field = record._fields[fname]
+                            val = field.convert_to_write(record[fname], record)
+                            if not record.id and field.type in ("one2many", "many2many")\
+                                and val == [(6, 0, origin[fname].ids)]:
+                                continue
```
Comment (rco-odoo) [inline odoo/fields.py:2916]: Add a comment to explain why.  I already forgot why we need this.

File: odoo/fields.py [hunk 21]
```diff
@@ -3707,6 +3734,7 @@ def apply_required(model, field_name):
     # At the time this function is called, the model's _fields may have been reset, although
     # the model's class is still the same. Retrieve the field to see whether the NOT NULL
     # constraint still applies
+    # VFE TODO postpone if field.compute ?
```
Comment (rco-odoo) [inline odoo/fields.py:3737]: No, you can remove that line.

File: odoo/modules/registry.py (hunk 12 above)
Comment (rco-odoo) [reply odoo/modules/registry.py:337]: Well, the tests pass without this hack being used, so you can remove it.

File: addons/sms/wizard/sms_composer.py (hunk 16 above)
Comment (Feyensv) [reply addons/sms/wizard/sms_composer.py:80]: It's a stored field on a wizard, the field should never have different values in cache.

File: odoo/fields.py (hunk 11 above)
Comment (Feyensv) [reply odoo/fields.py:2672]: Still needed, change merged with commit adding targeted tests

File: odoo/modules/registry.py (hunk 12 above)
Comment (Feyensv) [reply odoo/modules/registry.py:337]: I've added a "test" to highlight the problem, the error log is still there in the runbot ;)

File: odoo/addons/test_new_api_extension/models/test_fields.py [hunk 22]
```diff
@@ -0,0 +1,12 @@
+from odoo import api, fields, models
+from odoo.exceptions import ValidationError
+
+class DummyModel(models.Model):
+    _name = 'test_new_api.dummy'
```
Comment (rco-odoo) [inline odoo/addons/test_new_api_extension/models/test_fields.py:5]: This is no model extension!
```suggestion
    _inherit = 'test_new_api.dummy'
```

Comment (Feyensv) [reply odoo/addons/test_new_api_extension/models/test_fields.py:5]: Indeed :facepalm:

File: odoo/modules/registry.py (hunk 12 above)
Comment (rco-odoo) [reply odoo/modules/registry.py:337]: @Feyensv I just pushed a commit that fixes the source of the problem.

File: odoo/sql_db.py [hunk 23]
```diff
@@ -439,6 +439,8 @@ def commit(self):
         self.prerollback.clear()
         self.postrollback.clear()
         self.postcommit()
+        if hasattr(self, '_now'):
+            del self._now
```
Comment (Feyensv) [inline odoo/sql_db.py:443]: Forgot to verify if it was necessary (but it's not needed for the fixes).

File: odoo/models.py [hunk 24]
```diff
@@ -3712,13 +3814,30 @@ def create(self, vals_list):
...
+            vals = {key: val for key, val in vals.items() if key not in bad_names}
+            if self._log_access:
+                vals.setdefault('create_uid', self.env.uid)
+                vals.setdefault('create_date', self.env.cr.now())
+                vals.setdefault('write_uid', self.env.uid)
+                vals.setdefault('write_date', self.env.cr.now())
+
             # add missing defaults
-            vals = self._add_missing_default_values(vals)
+            vals = self._add_missing_default_values(vals, bad_names)
```
Comment (Feyensv) [inline odoo/models.py:3840]: Defaults should be added to values before vals cleanup.

**Comment by Feyensv** [comment]
Resurrected in #80449
